browser/src/actions/results.ts

```diff
@@ -20,247 +19,252 @@ export const notifyIsFetchingCommit = createAction<string>(Actions.IS_FETCHING_C
 export const fetchedAvatar = createAction<Avatar[]>(Actions.FETCHED_AVATARS);
 export const fetchedAuthors = createAction<ActionedUser[]>(Actions.FETCHED_AUTHORS);
 
-function getQueryUrl(store: RootState, baseUrl: string, args: string[] = []): string {
-    const id = store.settings.id || '';
-    const queryArgs = args.concat([`id=${encodeURIComponent(id)}`]);
-    const serverurl = window['server_url'];
-    return `${serverurl}${baseUrl}?${queryArgs.join('&')}`;
-}
-export const actionCommit = (logEntry: LogEntry, name: string = '', value: string = '') => {
-    // tslint:disable-next-line:no-any
-    return async (dispatch: Dispatch<any>, getState: () => RootState) => {
-        dispatch(notifyIsFetchingCommit(logEntry.hash.full));
-        const state = getState();
-        const url = getQueryUrl(state, `action/${name}`, [`value=${encodeURIComponent(value)}`]);
-        return axios.post(url, logEntry).then(result => {
-            switch (name) {
-                case 'reset_soft':
-                case 'reset_hard':
-                    dispatch(refresh());
-                    break;
-                case 'newtag':
-                    break;
-                case 'newbranch':
-                    dispatch(getBranches());    
-                    break;
-            }
-            dispatch(updateCommitInList(result.data as LogEntry));
-        });
+export namespace ResultActions {
+    export const commitsRendered = createAction<number>(Actions.COMMITS_RENDERED);
+
+    export const actionCommit = (logEntry: LogEntry, name: string = '', value: string = '') => {
+        // tslint:disable-next-line:no-any
+        return async (dispatch: Dispatch<any>, getState: () => RootState) => {
+            dispatch(notifyIsFetchingCommit(logEntry.hash.full));
+
+            const store = getState();
+
+            store.vscode.api.postMessage({
+                cmd: 'doAction',
+                args: { 
+                    ...store.settings,
```

Comment:
I would suggest adding something akin to a requestId to all messages sent.
This way when you get a response back u can look for messages with the same request and you can tie the callback to your original request.

Reply:
Note: This `requestId` can be added in the generic class that has the `<messageBus>post` 

E.g. crude implementation
```typescript
class MessageBus {
    private readonly callbacks = new Map<string, Function>();
    constructor(vscodeApi:...){
        window.addEventListener('message', this.onMessage.bind(this));
    }
    public async post<T>(message: string, args: any): Promise<T>{
        const requestId = uuid();
        const deferred = createDeferred();
        this.callbacks.set(requestId, deferred.Promise);
        this.vscode.api.postMessage(.....)
        return deferred.Promise;
    }
    protected onMessage(message:any){
         // Up to you how you structure the message coming in...
         this.callbacks.get(message.requestId)(message.payload);
    }
}

```

Reply:
I assume the MessageBus class should be available at almost all places but I think putting it into the redux store might be at the wrong place. Do you have any idea how to implement it as singleton alongside the redux store?

Reply:
Everywhere else use `post`

```typescript
import { post } from 'messagebus';

// in some thunk
.... post(...);
```